Implement better type for localize key string

steverep opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change

<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue or discussion
  in the additional information section.
-->
Instead of just requiring a string, this change requires that the string match a flattened key from the keys in `en.json`.  This also allows for autocompletion in editors.
## Type of change

<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example configuration

<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

```yaml

```

## Additional information

<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

steverep committed:
Implement better type for localize key string

Instead of just requiring a string, this change requires that the string match a flattened key from the keys in en.json.  This also allows for autocompletion in editors.

steverep committed:
Remove unused expected TS error directive

steverep committed:
Handle localize keys for supervisor

Localize function for supervisor takes keys relative to the "supervisor" top level key, so this must be handled with a separate function type using a generic.

steverep commented:
@bramkragten  it looks like this is catching a bunch of errors because of bad keys.  Can I get some assistance correcting them?

steverep committed:
Fix some bad translation keys for addons

steverep committed:
Fix some bad translation keys for pickers

bramkragten commented:
This is great!

Most of the issues are from dynamic keys, but I also see some issues with backend translations.

Not all translation keys are in the `en.json` we also get a bunch from the backend (integration specific stuff), that is not accounted for now in this code, like the `component` category.

We should exclude those top level categories from type checking probably.

bramkragten committed:
Update hassio-addon-config.ts

steverep committed:
Move translation dictionary type to main types and refactor localize function type for cleaner supervisor handling

steverep committed:
Fix some bad localize keys for hardware

steverep committed:
Add translation text for energy comparison

Text was taken from #12877 and added here to the correct file.

steverep committed:
Fix key type errors for addon info

steverep commented:
The backend translations are easy to include by pulling the keys from `src/translations/backend` so I pushed that out.  It didn't fix any errors by itself though. 😞 

I fixed a few more, but there are still 49 more bad static (i.e. non-template) keys.  You can grep the log for `"argument of type '\""` to filter them out.  I've attached a list of them with the extra error fluff removed.  I could definitely use a lot of help fixing them because in many cases it's not very clear what the author was going for.

[bad-keys.log](https://github.com/home-assistant/frontend/files/8997308/bad-keys.log)

For the errors in keys using template literals, I'll defer to you on the best strategy for this PR, but me trying to fix them all would probably take years without working full time on it 😓  I fixed one in addon info as an example, but I picked that one because it was easy.  Other random ones I examined get complicated quick, even deriving the `string` or `any` type from the HA web socket package.

I'd say an option is to assume many of them are safe for now and add `@ts-expect-error`, but I also noticed a few that will clearly produce bad keys too.

_I started this PR because I was working on something else to make a bunch of labels more accessible and got sick of having to manually type them. I did not realize the can of worms I was opening..._ 🪱

steverep committed:
Fix bad close keys for ZWaveJS

steverep committed:
Fix bad keys and type errors on device config page

bramkragten commented:
I will help you sorting this out, but it won't be before next week...

steverep commented:
@bramkragten  I have an important question.  I see a lot of the template literals look like:

```
this.hass.localize(key) || EnglishText;
```

I presume that's done because it's knowingly expecting `localize` to return an empty string for a bad key, and then use the English otherwise?

I'm concerned it's also trying to catch cases where the key exists, but the translation to a particular language may not?  That would make the fixes very problematic.

And an unrelated question: Is there a definitive list of all possible domains somewhere (e.g. string array or object with domain names as keys)?  The keys in the backend translations don't seem to include them all and I cannot find one elsewhere in the code.

bramkragten commented:
I don't think we should keep a fixed list of the backend domains/integrations, it is too dynamic.

The backend translations in the frontend are not up to date, they are requested from the backend and generated at runtime.

The usage of `this.hass.localize(key) || EnglishText;` should only be used as a fallback for when the key might not exist and localize will return an empty string, or for when localize and the translations are not loaded / initialized yet.

steverep committed:
Fix key type errors for triggers

Strategy is to assert the schema as const so its type is literals, then computeLabel types its schema from that. Other notes:
- Schema arrays must be typed as readonly
- Target computeLabel in ha-form cannot take any schema, but rather a much stronger type, so this should accept any
- Fixed other minor errors (e.g. missing imports)

steverep commented:
Okay to your last answer.  Got it.

I felt like a challenge, so I took on creating a strategy to fix the type errors involving computing labels for `ha-form`.  Take a look at my last commit and notes in the message when you get a chance.  Let me know what you think.  Note that it exposed a bug in the calendar trigger, further making the case for this change! 🛠️ 

And now this PR has officially gotten big.  Do you want to break it up somehow?  I've knocked off only ~50 of ~300 errors.

steverep committed:
Fix key type errors for wait and repeat actions

steverep committed:
Fix key type errors for conditions

steverep committed:
Fix key type errors for backup content

bramkragten commented on src/panels/config/automation/trigger/types/ha-automation-trigger-time_pattern.ts:
```diff
@@ -1,22 +1,22 @@
 import type { TriggerElement } from "../ha-automation-trigger-row";
 
-const SCHEMA: HaFormSchema[] = [
+const SCHEMA = [
```
I don't like that our `SCHEMA` is no longer typechecked now, I think that might be more important.

bramkragten replied:
Could be solved by adding:
```ts
const TYPED_SCHEMA: readonly Readonly<HaFormSchema>[] = SCHEMA;
```
But not really pretty 🤷‍♂️

steverep replied:
Yeah I agree.  I was thinking the same thing.  Theoretically though, this should be checked in the assignment to the `<ha-form>` property, but that's currently only set to warnings in the plugin:

```
"no-incompatible-type-binding": "warning",
```

And is lit even checked in the workflow?

bramkragten replied:
Yeah, it could be checked, but I'm afraid we would have to fix a LOT of other places then.

steverep replied:
Well, that's not really a bad thing.  Probably worth another branch to see just how bad it would be?  Most places I see that warning are pretty simple fixes involving `null` or `undefined`.  But that's coming from the guy who started this PR, so what do I know. 😇

steverep replied:
Actually, I just realized that some code I wrote to handle grid schemas might solve the problem:

```typescript
// Type utility to unionize a schema array by flattening any grid schemas
export type SchemaUnion<
  SchemaArray extends readonly HaFormSchema[],
  Schema = SchemaArray[number]
> = Schema extends HaFormGridSchema ? SchemaUnion<Schema["schema"]> : Schema;
```

If you always use this on the schema parameter for the callbacks, then it will be checked against the `extends readonly HaFormSchema[]` constraint, which is the type check you're after.

steverep replied:
Take a look at the latest commit for examples.  The calendar card schema was not actually being type checked, and the addition of `SchemaUnion` caught an error which I fixed.

bramkragten replied:
I would not be against creating a branch with:

`"no-incompatible-type-binding": "error"`

To see if we can fix those issues :-)

steverep committed:
Merge branch 'dev' into localize-key-type

steverep committed:
Fix key type errors for card editors (Round 1)

steverep committed:
Merge branch 'localize-key-type' of https://github.com/steverep/frontend into localize-key-type

steverep committed:
Fix key type errors for card editors (Round 2)

steverep committed:
Fix bad merge of duplicate energy compare key

steverep committed:
Fix key type errors for card editors (Round 3)

steverep committed:
Fix key type errors for card editors (Round 4)

steverep committed:
Update automation editor to use SchemaUnion

steverep committed:
Fix some key type errors in cards

steverep committed:
Fix key type errors for remaining form schemas

steverep committed:
Fix key type errors for blueprints

steverep committed:
Fix key type errors for disabled_by

steverep committed:
Fix key type errors for media player

steverep committed:
Fix key type errors for users

steverep committed:
Fix key type errors for cloud TTS

steverep committed:
Fix key type errors for user groups

steverep committed:
Grant localize key exception for component.*

steverep committed:
Fix key type errors in target picker

steverep committed:
Fix some key type errors in lovelace editors

steverep commented:
@bramkragten is this getting too big to review?  I don't want my effort going to waste by this sitting around forever.

I can temporarily add  `| string` to the the key, effectively disabling the type check and allowing this to be merged as is and using smaller PRs to get rid of a few errors at a time.

If need be, I can cherry pick apart this one to break it up too.  Let me know before I go any further.

steverep commented:
Nevermind... I convinced myself the cherry picking was worth it.  I can proceed once #13244 is merged.
